Assets/Scripts/Map/MapPatrollingGen/WatchmanRouteSolver.cs

```diff
@@ -0,0 +1,417 @@
+                }
+
+                var neighborsList = nearestNeighbors[start];
+                neighborsList.Add((end, distance));
+
+                // Sort by distance and keep only the top k nearest neighbors
+                neighborsList.Sort((a, b) => a.Item2.CompareTo(b.Item2));
+                if (neighborsList.Count > k)
+                {
+                    neighborsList.RemoveAt(neighborsList.Count - 1); // Remove the farthest neighbor
+                }
+            }
+
+            // Reverse nearest neighbor mapping
+            var reverseNearestNeighbors = new Dictionary<Vector2Int, List<Vector2Int>>();
+
+            foreach (var (start, neighborsList) in nearestNeighbors)
+            {
+                foreach (var (neighbor, _) in neighborsList)
+                {
+                    if (!reverseNearestNeighbors.ContainsKey(neighbor))
+                    {
+                        reverseNearestNeighbors[neighbor] = new List<Vector2Int>();
+                    }
+                    reverseNearestNeighbors[neighbor].Add(start);
+                }
+            }
+
+            return reverseNearestNeighbors;
+        }
+
+        // Depth-first search to find connected components
+        private static void DepthFirstSearch(Vertex vertex, HashSet<Vertex> visited, List<Vertex> component)
+        {
+            if (visited.Contains(vertex))
+            {
+                return;
+            }
+
+            visited.Add(vertex);
+            component.Add(vertex);
+
+            foreach (var neighbor in vertex.Neighbors)
+            {
+                if (!visited.Contains(neighbor))
+                {
+                    DepthFirstSearch(neighbor, visited, component);
+                }
+            }
+        }
+
+        // Merge islands (isolated connected verticies) recursively until all islands are connected
+        private static List<Vertex> MergeIslandsRecursively(
+            List<Vertex> currentIsland,
+            List<List<Vertex>> remainingIslands,
+            Dictionary<(Vector2Int, Vector2Int), int> distanceDict)
+        {
+            if (remainingIslands.Count == 0)
+            {
+                return currentIsland; // Base case: no more islands to merge
+            }
+
+            List<Vertex> closestIsland = null;
+            Vertex closestVertexInCurrent = null;
+            Vertex closestVertexInNew = null;
+            var minDistance = int.MaxValue;
+
+            // Find the closest island and vertices to merge
+            foreach (var island in remainingIslands)
+            {
+                foreach (var currentVertex in currentIsland)
+                {
+                    foreach (var newVertex in island)
+                    {
+                        if (distanceDict.TryGetValue((currentVertex.Position, newVertex.Position), out var distance) ||
+                            distanceDict.TryGetValue((newVertex.Position, currentVertex.Position), out distance))
```

Comment:
```suggestion
                        if (distanceDict.TryGetValue((currentVertex.Position, newVertex.Position), out var distance))
```

If we have pair (x, y) we have pair (y, x) right?